File: src/core/findings.js
const SEVERITY_ORDER = {
  error: 0,
  warn: 1,
  info: 2,
};

export function makeFinding(finding) {
  return {
    category: "general",
    detail: "",
    file: null,
    fixIds: [],
    fixable: false,
    hint: "",
    severity: "warn",
    ...finding,
  };
}

export function sortFindings(findings) {
  return findings.slice().sort((left, right) => {
    const severityDelta = SEVERITY_ORDER[left.severity] - SEVERITY_ORDER[right.severity];
    if (severityDelta !== 0) {
      return severityDelta;
    }

    const fileDelta = (left.file ?? "").localeCompare(right.file ?? "");
    if (fileDelta !== 0) {
      return fileDelta;
    }

    return left.title.localeCompare(right.title);
  });
}

export function uniqueFixes(fixes) {
  const seen = new Set();
  const unique = [];

  for (const fix of fixes) {
    if (seen.has(fix.id)) {
      continue;
    }

    seen.add(fix.id);
    unique.push(fix);
  }

  return unique;
}

export function summarizeFindings(findings, fixes, structure) {
  const blockingFindings = findings.filter((finding) => finding.severity === "error").length;
  const warningFindings = findings.filter((finding) => finding.severity === "warn").length;
  const infoFindings = findings.filter((finding) => finding.severity === "info").length;
  const fixableFindings = findings.filter((finding) => finding.fixable).length;

  let status = "clean";
  if (blockingFindings > 0) {
    status = "blocking issues";
  } else if (warningFindings > 0) {
    status = "attention needed";
  }

  return {
    status,
    totalFindings: findings.length,
    blockingFindings,
    warningFindings,
    infoFindings,
    fixableFindings,
    fixesAvailable: fixes.length,
    configFiles: structure.configFiles,
    packageCount: structure.packageCount,
    envDirectories: structure.envDirectories,
  };
}

export function serializeAuditResult(result) {
  return {
    rootDir: result.rootDir,
    summary: result.summary,
    structure: result.structure,
    findings: result.findings.map(serializeFinding),
    fixes: result.fixes.map((fix) => ({
      id: fix.id,
      title: fix.title,
      files: fix.files,
    })),
  };
}

function serializeFinding(finding) {
  return {
    id: finding.id,
    severity: finding.severity,
    category: finding.category,
    title: finding.title,
    detail: finding.detail,
    file: finding.file,
    hint: finding.hint,
    fixable: finding.fixable,
    fixIds: finding.fixIds,
  };
}

File: crates/maximus-core/src/findings.rs
use serde::Serialize;

use crate::models::{AuditResult, AuditSummary, Finding, FixPlan, Severity, StructureReport};
use crate::text_order::locale_compare_like;

pub const JSON_GENERATOR: &str = "maximus";
pub const JSON_SCHEMA_VERSION: &str = "1";

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FindingInput {
    pub id: String,
    pub title: String,
    pub category: Option<String>,
    pub detail: Option<String>,
    pub file: Option<std::path::PathBuf>,
    pub fix_ids: Vec<String>,
    pub fixable: bool,
    pub hint: Option<String>,
    pub severity: Option<Severity>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SerializableAuditResult {
    pub schema_version: &'static str,
    pub generator: &'static str,
    pub root_dir: std::path::PathBuf,
    pub summary: AuditSummary,
    pub structure: StructureReport,
    pub findings: Vec<SerializableFinding>,
    pub fixes: Vec<SerializableFixPlan>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SerializableFinding {
    pub id: String,
    pub severity: Severity,
    pub category: String,
    pub title: String,
    pub detail: String,
    pub file: Option<std::path::PathBuf>,
    pub hint: String,
    pub fixable: bool,
    pub fix_ids: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct SerializableFixPlan {
    pub id: String,
    pub title: String,
    pub files: Vec<std::path::PathBuf>,
}

pub fn make_finding(input: FindingInput) -> Finding {
    Finding {
        id: input.id,
        severity: input.severity.unwrap_or(Severity::Warn),
        category: input.category.unwrap_or_else(|| "general".to_string()),
        title: input.title,
        detail: input.detail.unwrap_or_default(),
        file: input.file,
        hint: input.hint.unwrap_or_default(),
        fixable: input.fixable,
        fix_ids: input.fix_ids,
    }
}

pub fn sort_findings(findings: &[Finding]) -> Vec<Finding> {
    let mut sorted = findings.to_vec();
    sorted.sort_by(|left, right| {
        let left_file = left
            .file
            .as_ref()
            .map(|value| value.to_string_lossy().to_string())
            .unwrap_or_default();
        let right_file = right
            .file
            .as_ref()
            .map(|value| value.to_string_lossy().to_string())
            .unwrap_or_default();

        severity_rank(&left.severity)
            .cmp(&severity_rank(&right.severity))
            .then_with(|| locale_compare_like(&left_file, &right_file))
            .then_with(|| locale_compare_like(&left.title, &right.title))
    });
    sorted
}

pub fn unique_fixes(fixes: &[FixPlan]) -> Vec<FixPlan> {
    let mut seen = std::collections::BTreeSet::new();
    let mut unique = Vec::new();

    for fix in fixes {
        if seen.insert(fix.id.clone()) {
            unique.push(fix.clone());
        }
    }

    unique
}

pub fn summarize_findings(
    findings: &[Finding],
    fixes: &[FixPlan],
    structure: &StructureReport,
) -> AuditSummary {
    summarize_findings_with_suppressed_by_config(findings, fixes, structure, 0)
}

pub fn summarize_findings_with_suppressed_by_config(
    findings: &[Finding],
    fixes: &[FixPlan],
    structure: &StructureReport,
    suppressed_by_config: usize,
) -> AuditSummary {
    let blocking_findings = findings
        .iter()
        .filter(|finding| finding.severity == Severity::Error)
        .count();
    let warning_findings = findings
        .iter()
        .filter(|finding| finding.severity == Severity::Warn)
        .count();
    let info_findings = findings
        .iter()
        .filter(|finding| finding.severity == Severity::Info)
        .count();
    let fixable_findings = findings.iter().filter(|finding| finding.fixable).count();

    let status = if blocking_findings > 0 {
        "blocking issues"
    } else if warning_findings > 0 {
        "attention needed"
    } else {
        "clean"
    };

    AuditSummary {
        status: status.to_string(),
        total_findings: findings.len(),
        blocking_findings,
        warning_findings,
        info_findings,
        fixable_findings,
        fixes_available: fixes.len(),
        suppressed_by_config,
        config_files: structure.config_files,
        package_count: structure.package_count,
        env_directories: structure.env_directories,
    }
}

pub fn serialize_audit_result(result: &AuditResult) -> SerializableAuditResult {
    SerializableAuditResult {
        schema_version: JSON_SCHEMA_VERSION,
        generator: JSON_GENERATOR,
        root_dir: result.root_dir.clone(),
        summary: result.summary.clone(),
        structure: result.structure.clone(),
        findings: result
            .findings
            .iter()
            .map(|finding| SerializableFinding {
                id: finding.id.clone(),
                severity: finding.severity.clone(),
                category: finding.category.clone(),
                title: finding.title.clone(),
                detail: finding.detail.clone(),
                file: finding.file.clone(),
                hint: finding.hint.clone(),
                fixable: finding.fixable,
                fix_ids: finding.fix_ids.clone(),
            })
            .collect(),
        fixes: result
            .fixes
            .iter()
            .map(|fix| SerializableFixPlan {
                id: fix.id.clone(),
                title: fix.title.clone(),
                files: fix.files.clone(),
            })
            .collect(),
    }
}

fn severity_rank(severity: &Severity) -> usize {
    match severity {
        Severity::Error => 0,
        Severity::Warn => 1,
        Severity::Info => 2,
    }
}
